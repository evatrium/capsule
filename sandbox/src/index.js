let log_this = '';
const log = (it) => {
    let val;
    if (typeof it !== 'string') {
        val = JSON.stringify(it, null, 4);
    } else {
        val = it;
    }
    log_this = '\n' + log_this + '\n' + val + '\n';
};


export const tester = () => {
    return new Promise((resolve, reject) => {
        console.log(log_this);

        console.log('working');


        resolve()
    })
};