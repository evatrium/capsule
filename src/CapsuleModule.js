import React, {Component} from "react";

class Root extends Component {
    render() {
        return (
            <React.Fragment>
                {this.props.children}
            </React.Fragment>
        )
    }
}

export const CapsuleModule = ({createStore, actionsCreator, classNames, asyncHOC, connect, Provider, jssProcessor, events}) => {

    const store = createStore();
    const cn = classNames;
    let collective = {};
    let actions = {};
    let theme = {};
    let propsFromProvider = {};
    let logicFromProvider = {};
    let _jssProcessor = jssProcessor;
    let stylesHOC;
    let RootComponent;

    const CapsuleProvider = (config = {}) => {

        theme = config.theme || {};
        propsFromProvider = config.toProps || {};
        logicFromProvider = config.toLogic || {};
        _jssProcessor = config.stylesHOC ? false : config.jssProcessor || jssProcessor;
        stylesHOC = config.stylesHOC;
        RootComponent = config.RootComponent || Root;
        collective = {...collective, ...config.toCollective};


        return (Child = null) => {

            const AC = asyncHOC && config.loadApp ? asyncHOC(config.loadApp) : null;

            const App = AC || Child;

            return () => (
                <Provider
                    store={store}
                    collective={{getCollective: () => collective}}
                    actions={{getActions: () => actions}}>
                    <RootComponent>
                        <App/>
                    </RootComponent>
                </Provider>
            );
        }
    };

    const makeActionsAndAttachState = (initialState, actionsName) => {
        if (!initialState || !actionsName) return;
        store.attachState({[actionsName]: initialState});
        actions = actionsCreator(actionsName, initialState, store, actions);
    };

    const makeLogic = (logic) => {
        if (!logic) return {};
        return logic({
            ...logicFromProvider,
            events,
            actions,
            store,
            collective: () => collective
        })
    };

    const updateCollective = (logicName, returnLogic) => {
        if (!logicName) return;
        collective = {
            ...collective,
            [logicName]: returnLogic
        }
    };

    const makeClassesFromStyles = (styles) => {
        if (!styles || !_jssProcessor) return null;
        return _jssProcessor(typeof styles === 'function' ? styles(theme) : styles);
    };

    const Capsule = (c = {}) => {

        let {
            name,
            styles,
            initialState,
            actionsName,
            mapState,
            mapActions,
            logicName,
            logic,
            mapLogic,
        } = c;

        if (name) {
            if (!actionsName) actionsName = name;
            if (!logicName) logicName = name;
        }

        makeActionsAndAttachState(initialState, actionsName);

        let returnLogic = makeLogic(logic);

        updateCollective(logicName, returnLogic);

        let classes = makeClassesFromStyles(styles);

        return (Child) => {
            if (!Child) return returnLogic;

            if (styles && stylesHOC) {
                Child = stylesHOC(styles)(Child);
            }

            return connect({
                mapState,
                mapActions,
                mapLogic,
                additionalProps: {
                    cn,
                    theme,
                    ...propsFromProvider,
                    classes
                }
            })(Child)
        };
    };

    return {
        CapsuleProvider,
        Capsule,
    }

};
