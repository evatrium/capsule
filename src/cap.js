import {Capsule} from "./index";


const user = Capsule('user', {
    active: false,
    first_name: '',
    last_name: '',
    src: ''
});

const onUserStateChange = ({active}) => {
    user.merge({active});
};

const Avatar = user(({active, first_name, last_name, src}) => {

    return (
        <UserBox>
            <Img src={src}/>
            <Typo text={first_name + ' ' + last_name}/>
        </UserBox>
    );
});

