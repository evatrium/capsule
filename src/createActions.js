import {basicNestedCopy} from "./utils";

export const createActions = (actionsName, initialState, store, actions, noAdditionalImmutability) => {
    initialState = basicNestedCopy(initialState, noAdditionalImmutability);
    //set up the placeholders for all the actions on the namespace
    actions[actionsName] = {set: {}, get: {}, toggle: {}, update: () => null, merge: () => null};

    actions[actionsName].update = (cb, skipCopy) => {
        let state = store.getState();
        // pass the namespace state to the callback,
        // and assign the returned value to the namespace
        store.setState({
            ...state,
            [actionsName]: basicNestedCopy(cb({...state[actionsName]}), noAdditionalImmutability || skipCopy)
        });
    };

    actions[actionsName].merge = (propsToMerge, skipCopy) => {
        let state = store.getState();
        //spread the propsToMerge and assign to the namespace
        store.setState({
            ...state,
            [actionsName]: {
                ...state[actionsName],
                ...basicNestedCopy(propsToMerge, noAdditionalImmutability || skipCopy)
            }
        });
    };

    //return the namespace state
    actions[actionsName].getState = (skipCopy) =>
        basicNestedCopy(store.getState()[actionsName], noAdditionalImmutability || skipCopy);

    //for every key in the state
    //set the value as the function name of the action type
    //{foo:1} ... namespace.set.foo(2);
    Object.keys(initialState).forEach(key => {

        //set a single value onto the namespace
        actions[actionsName].set[key] = (newValue, skipCopy) => {
            let state = store.getState();
            store.setState({
                ...state,
                [actionsName]: {
                    ...state[actionsName],
                    [key]: basicNestedCopy(newValue, noAdditionalImmutability || skipCopy)
                }
            });
        };

        //get a single value from the namespace
        actions[actionsName].get[key] = (skipCopy) => {
            return basicNestedCopy(store.getState()[actionsName][key], noAdditionalImmutability || skipCopy);
        };

        //toggle a single value on a namespace to its inverse
        actions[actionsName].toggle[key] = () => {
            actions[actionsName].set[key](!actions[actionsName].get[key]())
        };
    });

    return actions;
};
