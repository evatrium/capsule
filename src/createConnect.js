import React, {Component, createElement} from 'react';
import {convertArguments} from "./utils";

/**
 * selects the values off of the state
 * accepts a function to map props like this:
 * (state) => ({mySingleProp: state.namespace.prop});
 * or with comma separated values on an object like this:
 * {namespace:'prop,someOtherProp', someOtherNameSpace: 'foo,bar'}
 * @param {function|Object} mapper - plucks the keys off of the state
 * @param {Object} source - the state
 * @returns {Object} - the selected state
 */
export const mapIt = (mapper, source) => {
    if (typeof mapper === 'function') return mapper(source);
    if (typeof mapper === 'object') {
        let selected = {};
        let keys = Object.keys(mapper);
        for (let i = 0; i < keys.length; i++) {
            const namespace = keys[i];
            if (!source[namespace]) break;
            let props = mapper[namespace];
            if (typeof props === 'string') props = props.split(/\s*,\s*/);
            for (let i = 0; i < props.length; i++) selected[props[i]] = source[namespace][props[i]];
        }
        return selected;
    }
    return {};
};

/*** Wire a component up to the store. Passes state as props, re-renders on change.
 * inspiration: https://github.com/developit/unistore/blob/master/src/integrations/react.js
 * @param CapsuleContext
 * @returns {function}*/
export const createConnect = (CapsuleContext) => (...args) => Child => {

    const {mapState, mapActions, mapLogic} = convertArguments(...args);

    function Connected(props, context) {
        Component.call(this, props, context);
        context = context || {};
        const {store, getCollective, getActions} = context;

        let mappedProps = {}, state = {},
            getState = () => store && store.getState ? store.getState() : {},
            update = () => {
            }, sub, unsub;

        //map the collections to props
        if (mapLogic && getCollective) mappedProps = mapIt(mapLogic, getCollective());
        if (mapActions && getActions) mappedProps = {...mappedProps, ...mapIt(mapActions, getActions())};
        if (mapState) { //skip if there is no mapState in case we are just mapping logic or actions
            state = mapIt(mapState, getState()); //first run-through, map the state
            update = () => {
                let mapped = mapIt(mapState, getState()); //get the plucked state values onto an object
                for (let i in mapped) if (mapped[i] !== state[i]) {//and compare each value with the last
                    state = mapped; // if there is a change then update the state
                    //we can dip out at this point. check if mounted so we dont update after unmounting
                    return this.mounted && this.forceUpdate();//simulate setState on the wrapped component
                }// if there aren't any updates then check to see if any new properties have been added
                for (let i in state) if (!(i in mapped)) {
                    state = mapped;
                    return this.mounted && this.forceUpdate();
                }
            }; //subscribe the update function to be called whenever store.setState is called
            sub = () => store && store.subscribe(update);
            unsub = () => store && store.unsubscribe(update);
        }

        this.componentDidMount = () => {
            this.mounted = true;
            update(); //need to call update here
            //in case the component (that this wraps) triggers an update
            //prior to subscribing to the store.
            sub && sub();
        };
        this.componentWillUnmount = () => {
            this.mounted = false;
            unsub && unsub();
        };
        this.render = () =>
            createElement(Child, {...mappedProps, ...state, ...this.props});
    }

    const ConnectedCapsule = (Connected.prototype = new Component()).constructor = Connected;
    ConnectedCapsule.contextType = CapsuleContext;
    return ConnectedCapsule;
};
