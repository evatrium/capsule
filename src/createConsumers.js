import {createElement, useContext, useReducer, useRef, useMemo, useEffect} from 'react';

export const mapIt = (mapper, source) => {
    if (typeof mapper === 'function') return mapper(source);
    if (typeof mapper === 'object') {
        let selected = {};
        let keys = Object.keys(mapper);
        for (let i = 0; i < keys.length; i++) {
            const namespace = keys[i];
            if (!source[namespace]) break;
            let props = mapper[namespace];
            if (typeof props === 'string') props = props.split(/\s*,\s*/);
            for (let i = 0; i < props.length; i++) selected[props[i]] = source[namespace][props[i]];
        }
        return selected;
    }
    return {};
};

export const createConsumers = (CapsuleContext) => {

    const useForceUpdate = () => useReducer(state => !state, false)[1];

    const useCapsule = ({mapState, mapActions, mapLogic}) => {
        const {store, getCollective, getActions} = useContext(CapsuleContext);
        let forceUpdate = useForceUpdate(),
            ignoreForce = useRef(false),
            hasMounted = useRef(false),
            update = useRef(() => null),
            state = useRef(store.getState()),
            mappedProps = useRef({});

        mappedProps.current = useMemo(() => {
            let m_props = {};
            if (mapLogic && getCollective) m_props = mapIt(mapLogic, getCollective());
            if (mapActions && getActions) m_props = {...m_props, ...mapIt(mapActions, getActions())};
            return m_props;

        }, []);

        const updater = useMemo(() => (_, initial) => {
            let mapped = mapIt(mapState, store.getState());
            for (let i in mapped) if (mapped[i] !== state.current[i]) {
                state.current = mapped;
                return !ignoreForce.current && !initial && forceUpdate()
            }
            for (let i in state.current) if (!(i in mapped)) {
                state.current = mapped;
                return !ignoreForce.current && !initial && forceUpdate()
            }

        }, []);

        if (!hasMounted.current && mapState) updater(null, true);

        useEffect(() => {
            hasMounted.current = true;
            if (mapState) {
                update.current = updater;
                update.current();
            }
            const unsub = store.subscribe(update.current);
            return () => {
                ignoreForce.current = true;
                unsub();
            }

        }, []);

        return [state.current, mappedProps.current]
    };

    const connectCapsule = ({mapState, mapActions, mapLogic}) => Child => props => {
        const [state, mappedProps] = useCapsule({mapState, mapActions, mapLogic});
        return createElement(Child, {...mappedProps, ...state, ...props});
    };

    return {
        useCapsule,
        connectCapsule,
    }
};
