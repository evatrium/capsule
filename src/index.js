import {createCapsule} from "./createCapsule";
import {createRouting} from "./createRouting";

const {
    CapsuleProvider,
    Capsule,
    events,
    useCapsule,
    connectCapsule
} = createCapsule();

const {
    routing,
    Router,
    Linkage,
    pathSwitch
} = createRouting(Capsule);

export {
    CapsuleProvider,
    Capsule,
    events,
    useCapsule,
    connectCapsule,
    routing,
    Router,
    Linkage,
    pathSwitch,

};

