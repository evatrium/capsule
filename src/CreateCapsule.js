import {createStore, createActions, connect, Provider} from "@iosio/store";
import {classNames} from "@iosio/utils/lib/classNames";
import {lazyLoader } from "./lazyLoader";
import {CapsuleModule} from "./CapsuleModule";
import {JSS_nano} from "@iosio/jss_nano";
import {Eventer} from '@iosio/utils/lib/eventer';
import {createRouting} from "./routing";

export const CreateCapsule = () => {

    const events = Eventer();

    const jss_nano = JSS_nano(process.env.NODE_ENV === 'development');

    const {CapsuleProvider, Capsule} = CapsuleModule({
        createStore,
        connect,
        Provider,
        actionsCreator: createActions,
        classNames,
        asyncHOC: lazyLoader,
        jssProcessor: jss_nano,
        events
    });

    const {
        routing,
        Router,
        WithRouting,
        Linkage
    } = createRouting(Capsule);

    return {
        CapsuleProvider,
        Capsule,
        events,
        lazyLoader,
        routing,
        Router,
        Linkage,
        WithRouting,
        jss_nano
    }

};
