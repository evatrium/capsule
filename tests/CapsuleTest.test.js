import React from 'react';
import {CreateCapsule} from "../src/CreateCapsule";
import renderer from "react-test-renderer";
import {render, fireEvent, cleanup, queryByAttribute} from 'react-testing-library';

import {getById} from "./_test_utils";
import 'jest-dom/extend-expect'

const Printer = (props) => {
    return (
        <div>
            {JSON.stringify(props, null, 4)}
        </div>
    );
};


it(`
provides the expected arguments to logic: actions, store, collective, events.

routing actions and logic are provided to collective by default

returns logic when nothing is passed to the second curried function call

can map state, logic and actions to a component 

component updates with the correct values when logic and action functions are fired 

    `, () => {


    const {Capsule, CapsuleProvider} = CreateCapsule();


    const myStateLogic = Capsule({
        name: 'myCap1',
        initialState: {
            count: 0,
            value: 'initial'
        },
        logic: (logicArguments) => {

            expect(logicArguments).toMatchObject({
                actions: {
                    myCap1: {
                        set: {
                            count: expect.any(Function),
                            value: expect.any(Function),
                        },
                        get: {
                            count: expect.any(Function),
                            value: expect.any(Function),
                        },
                        toggle: {
                            count: expect.any(Function),
                            value: expect.any(Function),
                        },
                        update: expect.any(Function),
                        merge: expect.any(Function),
                        getState: expect.any(Function)
                    },
                },
                store: {
                    setState: expect.any(Function),
                    getState: expect.any(Function),
                    subscribe: expect.any(Function),
                    unsubscribe: expect.any(Function),
                    attachState: expect.any(Function)
                },
                events: {
                    on: expect.any(Function),
                    off: expect.any(Function),
                    once: expect.any(Function),
                    emit: expect.any(Function),
                    destroy: expect.any(Function)
                },
                collective: expect.any(Function)
            });


            return {
                logicTest: (val) => {
                    logicArguments.actions.myCap1.set.value(val)
                }
            }

        }

    })();

    expect(myStateLogic).toMatchObject({
        logicTest: expect.any(Function)
    });


    const MyComponent = (props) => {

        expect(props).toMatchObject({
            set: {
                count: expect.any(Function),
                value: expect.any(Function)
            },
            logicTest: expect.any(Function),
            count: expect.any(Number),
            value: expect.any(String),
            cn: expect.any(Function),
            theme: {},
            classes: null
        });

        return (
            <div>
                <button id={'setValue'} onClick={() => {
                    props.logicTest('updated')
                }}>
                    {props.value}
                </button>

                <button id={'setCount'} onClick={() => {
                    props.set.count(props.count + 1)
                }}>
                    {props.count}
                </button>


                {JSON.stringify(props, null, 4)}
            </div>
        );
    };


    const MyStateLogicComponent = Capsule({
        mapLogic: {myCap1: 'logicTest'},
        mapState: {myCap1: 'count,value'},
        //or optionally use function
        mapActions: ({myCap1}) => ({set: myCap1.set})
    })(MyComponent);


    const App = CapsuleProvider()(({children}) => {

        return (
            <div>
                <MyStateLogicComponent/>
                Hello
            </div>
        )
    });


    const {container} = render(<App/>);

    const setValueBtn = () => getById(container, 'setValue');
    const setCountBtn = () => getById(container, 'setCount');

    expect(setValueBtn()).toHaveTextContent('initial');

    expect(setCountBtn()).toHaveTextContent('0');

    fireEvent.click(setValueBtn());

    expect(setValueBtn()).toHaveTextContent('updated');

    fireEvent.click(setCountBtn());

    expect(setCountBtn()).toHaveTextContent('1');

    fireEvent.click(setCountBtn());

    expect(setCountBtn()).toHaveTextContent('2');

    expect(setValueBtn()).toHaveTextContent('updated');


    const tree = renderer.create(<App/>);

    expect(tree).toMatchSnapshot()

});
