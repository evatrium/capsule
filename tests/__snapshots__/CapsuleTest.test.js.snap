// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`
provides the expected arguments to logic: actions, store, collective, events.

routing actions and logic are provided to collective by default

returns logic when nothing is passed to the second curried function call

can map state, logic and actions to a component 

component updates with the correct values when logic and action functions are fired 

     1`] = `
Array [
  <div>
    <button
      id="setValue"
      onClick={[Function]}
    >
      updated
    </button>
    <button
      id="setCount"
      onClick={[Function]}
    >
      2
    </button>
    {
    "set": {},
    "count": 2,
    "value": "updated"
}
  </div>,
  "Hello",
]
`;
